Update Migrations Guide (#15130)
pimcore/pimcore: doc/23_Installation_and_Upgrade/07_Updating_Pimcore/12_V10_to_V11.md

# Upgrading Pimcore from Version 10.x to Version 11

> ## Tasks to do prior the Update

Reviewer: ```suggestion
## Tasks to Do Prior the Update
```

### Migrate System Settings

> System Settings are using LocationAwareConfigRepository in Pimcore 11. To keep your specific settings, please follow steps below: 

Reviewer: ```suggestion
System settings are using LocationAwareConfigRepository in Pimcore 11. To keep your specific settings, please follow the steps below: 
```

#### Appearance & Branding 

> Appearance & Branding settings will be separated from the System settings in Pimcore 11 and stored in `var/config/admin_system_settings/admin_system_settings.yaml` by default. Please copy every relevant settings from `var/config/system.yaml` into `var/config/admin_system_settings/admin_system_settings.yaml` to keep your settings. 

Reviewer: ```suggestion
Appearance & Branding settings will be separated from the system settings in Pimcore 11 and stored in `var/config/admin_system_settings/admin_system_settings.yaml` by default. Please copy every relevant settings from `var/config/system.yaml` into `var/config/admin_system_settings/admin_system_settings.yaml` to keep your settings. 
```

> To save these settings into the settings store, you can use following configuration:

Reviewer: ```suggestion
To save these settings into the settings store, you can use the following configuration:
```

```yaml
pimcore_admin:
    config_location:
        admin_system_settings:
            write_target:
                type: 'settings-store'
            read_target:
                type: 'settings-store'
```

#### System Settings

Please copy all relevant settings from `var/config/system.yaml` into `var/config/system_settings/system_settings.yaml`.

> To save system settings into the settings store, please add following to your configuration:

Reviewer: ```suggestion
To save system settings into the settings store, please add the following to your configuration:
```

```yaml
pimcore:
    config_location:
        system_settings:
            write_target:
                type: 'settings-store'
            read_target:
                type: 'settings-store'
```

### Configuration Storage Settings

> As some features were moved to separate bundles, their configuration storage settings might need to be moved. This is a two step process 

Reviewer: ```suggestion
As some features were moved to separate bundles, their configuration storage settings might need to be moved. This is a two steps process:
```

> 1) Remove settings prior to update 

Reviewer: ```suggestion
1) Remove settings before the update.
```

> 2) Readd settings after extracted bundles are installed

Reviewer: ```suggestion
2) Re-Add settings after the extracted bundles installation.
```

> Relevant configs are: 

Reviewer: ```suggestion
The relevant configs are: 
```

```yaml
# pimcore_custom_reports:
#  config_location:
#    custom_reports:
#      write_target:
#        type: 'settings-store'

# pimcore_web_to_print:
#  config_location:
#    web_to_print:
#      write_target:
#        type: 'settings-store'

# pimcore_static_routes:
#  config_location:
#    staticroutes:
#      write_target:
#        type: 'settings-store'
```

### Web2Print

> If you are using Web2Print functionality set the flag "Enable Web2Print documents in default documents view" to true to ensure the PimcoreWebToPrintBundle gets installed by default.

Reviewer: ```suggestion
If you are using the Web2Print functionality, set the flag "Enable Web2Print documents in default documents view" to true to ensure the PimcoreWebToPrintBundle gets installed by default.
```

> If you are not using Web2Print functionality set the flag to false.

Reviewer: ```suggestion
If you are not using Web2Print functionality, set the flag to false.
```

## Update Pimcore via Composer

Update Pimcore to Pimcore 11 via composer statement `composer require -W pimcore/pimcore:^11.0 pimcore/admin-ui-classic-bundle`. 

> Depending on your setup, it might be necessary to include additional dependencies in the update statement and/or cleanup `composer.lock` file and `vendor` folder.

Reviewer: ```suggestion
Depending on your setup, it might be necessary to include additional dependencies in the update statement and/or to clean up the `composer.lock` file and the `vendor` folder.
```

> After composer tasks like `bin/console assets:install` might fail due to additional adaptions necessary before symfony container can be built again. See following tasks. 

Reviewer: ```suggestion
After composer, some tasks like `bin/console assets:install` might fail due to necessary additional adaptions before the Symfony container can be built again. For more info, see the following [Tasks](#Tasks to Do After the Update) section.
```

> ## Tasks to do after the Update

Reviewer: ```suggestion
## Tasks to Do After the Update
```

### Admin Bundle
Register the admin-ui-classic-bundle in Kernel (and adapt method signature if necessary): 
`src/Kernel.php`
```php
public function registerBundlesToCollection(BundleCollection $collection): void
{
    // pimcore bundles
    $collection->addBundle(new \Pimcore\Bundle\AdminBundle\PimcoreAdminBundle(), 60);
}
```

> ### Some code changes

Reviewer: ```suggestion
### Some Code Changes
```

> - Update `config/packages/security.yaml` to latest format. As reference use [security.yaml](https://github.com/pimcore/skeleton/blob/11.x/config/packages/security.yaml) of our skeleton. 

Reviewer: ```suggestion
- Update `config/packages/security.yaml` to the latest format. As reference use [security.yaml](https://github.com/pimcore/skeleton/blob/11.x/config/packages/security.yaml) of our skeleton. 
```

> - Update references to `@PimcoreCoreBundle/Resources/config/...` to `@PimcoreCoreBundle/config/...`

Reviewer: ```suggestion
- Update references from `@PimcoreCoreBundle/Resources/config/...` to `@PimcoreCoreBundle/config/...`
```

>   - especially in `config/routes.yaml` (probably you also need to change `routing.yml` to `routing.yaml` there)  

Reviewer: ```suggestion
  - especially in `config/routes.yaml` (you probably also need to change `routing.yml` to `routing.yaml` there). 
```

> ### Install extracted bundles via composer if necessary

Reviewer: ```suggestion
### Install Extracted Bundles via Composer if Necessary
```

> Add extracted bundles (such as PimcoreEcommerceFrameworkBundle or PimcorePersonalizationBundle) if necessary via Composer. 

Reviewer: ```suggestion
Add the extracted bundles (such as PimcoreEcommerceFrameworkBundle or PimcorePersonalizationBundle) via composer if necessary. 
```

```bash 
composer require pimcore/system-info-bundle pimcore/file-explorer-bundle pimcore/personalization-bundle pimcore/google-marketing-bundle pimcore/web-to-print-bundle pimcore/ecommerce-framework-bundle pimcore/newsletter-bundle
```

Skip bundles as per your requirements. 

> ### Activate extracted bundles in `config/bundles.php` if necessary

Reviewer: ```suggestion
### Activate Extracted Bundles in `config/bundles.php` if Necessary
```

> Update `config/bundles.php` and add all needed extracted bundles to the array: 

Reviewer: ```suggestion
Update `config/bundles.php` and add all the needed extracted bundles to the array: 
```

  \Pimcore\Bundle\SimpleBackendSearchBundle\PimcoreSimpleBackendSearchBundle::class => ['all' => true],
  \Pimcore\Bundle\CustomReportsBundle\PimcoreCustomReportsBundle::class => ['all' => true],
  \Pimcore\Bundle\GoogleMarketingBundle\PimcoreGoogleMarketingBundle::class => ['all' => true],
  \Pimcore\Bundle\ApplicationLoggerBundle\PimcoreApplicationLoggerBundle::class => ['all' => true],
  \Pimcore\Bundle\WebToPrintBundle\PimcoreWebToPrintBundle::class => ['all' => true], 
  \Pimcore\Bundle\TinymceBundle\PimcoreTinymceBundle::class => ['all' => true],
  \Pimcore\Bundle\StaticRoutesBundle\PimcoreStaticRoutesBundle::class => ['all' => true],
  \Pimcore\Bundle\NewsletterBundle\PimcoreNewsletterBundle::class => ['all' => true],
  \Pimcore\Bundle\WordExportBundle\PimcoreWordExportBundle::class => ['all' => true],
  \Pimcore\Bundle\XliffBundle\PimcoreXliffBundle::class => ['all' => true],
  \Pimcore\Bundle\FileExplorerBundle\PimcoreFileExplorerBundle::class => ['all' => true],
  \Pimcore\Bundle\SystemInfoBundle\PimcoreSystemInfoBundle::class => ['all' => true],
```

Remove bundles as per your requirements. 

### Configuration Storage Settings - Part 2

> Reactivate configuration storage settings for extrated bundles as needed: 

Reviewer: ```suggestion
Reactivate the configuration storage settings for the extracted bundles as needed: 
```

```yaml
#custom_reports
pimcore_custom_reports:
  config_location:
    custom_reports:
      ...

#web_to_print
pimcore_web_to_print:
  config_location:
    web_to_print:
       ...

#static_routes
pimcore_static_routes:
  config_location:
    staticroutes:
       ...
```

### Migrations

> As soon as the Symfony container can be built again, execute all migrations. 

Reviewer: ```suggestion
As soon as the Symfony container can be built again, you can execute all the migrations:
```

```bash
bin/console doctrine:migrations:migrate`
```

### Web2Print

> If you do not use Web2Print functionality, then please make sure to run the following command and queries to clean up your system.

Reviewer: ```suggestion
If you do not use the Web2Print functionality, then please make sure to run the following command and queries to clean up your system:
```

```bash
bin/console pimcore:documents:cleanup printpage printcontainer
```

```sql
DELETE FROM `users_permission_definitions` WHERE `key` = 'web2print_settings';
UPDATE `users` SET `permissions`=REGEXP_REPLACE(`permissions`, '(?:^|,)web2print_settings(?:$|,)', '') WHERE `permissions` REGEXP '(?:^|,)web2print_settings(?:$|,)';
```

### Newsletters

> If you do not use Newsletter functionality, then please make sure to run the following command to clean up your system.

Reviewer: ```suggestion
If you do not use the Newsletter functionality, then please make sure to run the following command to clean up your system:
```

```

After composer tasks like `bin/console assets:install` might fail due to additional adaptions necessary before symfony container can be built again. See following tasks. 

## Tasks to do after the Update

### Admin Bundle
Register the admin-ui-classic-bundle in Kernel (and adapt method signature if necessary): 
`src/Kernel.php`
```php
public function registerBundlesToCollection(BundleCollection $collection): void
{
    // pimcore bundles
    $collection->addBundle(new \Pimcore\Bundle\AdminBundle\PimcoreAdminBundle(), 60);
}
```

### Some code changes

- Update `config/packages/security.yaml` to latest format. As reference use [security.yaml](https://github.com/pimcore/skeleton/blob/11.x/config/packages/security.yaml) of our skeleton. 
- Update references to `@PimcoreCoreBundle/Resources/config/...` to `@PimcoreCoreBundle/config/...`
  - especially in `config/routes.yaml` (probably you also need to change `routing.yml` to `routing.yaml` there)  

### Install extracted bundles via composer if necessary

> Add extrated bundles if necessary via composer. 

Reviewer 2: ```suggestion
Add extracted bundles (such as PimcoreEcommerceFrameworkBundle or PimcorePersonalizationBundle) if necessary via Composer. 
```

  \Pimcore\Bundle\WebToPrintBundle\PimcoreWebToPrintBundle::class => ['all' => true], 
  \Pimcore\Bundle\TinymceBundle\PimcoreTinymceBundle::class => ['all' => true],
  \Pimcore\Bundle\StaticRoutesBundle\PimcoreStaticRoutesBundle::class => ['all' => true],
  \Pimcore\Bundle\NewsletterBundle\PimcoreNewsletterBundle::class => ['all' => true],
  \Pimcore\Bundle\WordExportBundle\PimcoreWordExportBundle::class => ['all' => true],
  \Pimcore\Bundle\XliffBundle\PimcoreXliffBundle::class => ['all' => true],
  \Pimcore\Bundle\FileExplorerBundle\PimcoreFileExplorerBundle::class => ['all' => true],
  \Pimcore\Bundle\SystemInfoBundle\PimcoreSystemInfoBundle::class => ['all' => true],
```

Remove bundles as per your requirements. 

### Configuration Storage Settings - Part 2
Reactivate configuration storage settings for extrated bundles as needed: 
```yaml
#custom_reports
pimcore_custom_reports:
  config_location:
    custom_reports:
      ...

#web_to_print
pimcore_web_to_print:
  config_location:
    web_to_print:
       ...

#static_routes
pimcore_static_routes:
  config_location:
    staticroutes:
       ...
```

### Migrations

> As soon as symfony container can be built again, execute all migrations. 

Reviewer 2: ```suggestion
As soon as the Symfony container can be built again, execute all migrations. 
```

```bash
bin/console doctrine:migration:exec 'Pimcore\Bundle\CoreBundle\Migrations\Version20230412105530'
```

### Additional Cleanups 

> - Remove `var/config/system.yaml` after all settings are migrated. 

Reviewer: ```suggestion
- Remove `var/config/system.yaml` after migrating all settings. 
```
